File: src/ai/tools/host/fuzzy-edit.ts
// ---------------------------------------------------------------------------
// Fuzzy Edit — 5-strategy matching for file edits
//
// Strategies tried in order:
// 1. Exact match
// 2. Line-trimmed (trim each line, compare)
// 3. Whitespace-normalized (collapse internal whitespace)
// 4. Indentation-flexible (strip common indent, re-indent replacement)
// 5. Block-anchor + Levenshtein (match first/last line, 30% tolerance)
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface FuzzyMatchResult {
	readonly replaced: string;
	readonly strategy: string;
}

// ---------------------------------------------------------------------------
// Levenshtein distance
// ---------------------------------------------------------------------------

function levenshtein(a: string, b: string): number {
	const m = a.length;
	const n = b.length;

	if (m === 0) return n;
	if (n === 0) return m;

	// Use two rows instead of full matrix
	let prev = new Uint32Array(n + 1);
	let curr = new Uint32Array(n + 1);

	for (let j = 0; j <= n; j++) prev[j] = j;

	for (let i = 1; i <= m; i++) {
		curr[0] = i;
		for (let j = 1; j <= n; j++) {
			const cost = a[i - 1] === b[j - 1] ? 0 : 1;
			curr[j] = Math.min(
				prev[j] + 1, // deletion
				curr[j - 1] + 1, // insertion
				prev[j - 1] + cost, // substitution
			);
		}
		[prev, curr] = [curr, prev];
	}

	return prev[n];
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function splitLines(text: string): string[] {
	return text.split('\n');
}

function getCommonIndent(lines: readonly string[]): string {
	const nonEmpty = lines.filter((l) => l.trim().length > 0);
	if (nonEmpty.length === 0) return '';

	let indent: string | undefined;
	for (const line of nonEmpty) {
		const match = line.match(/^(\s*)/);
		const lineIndent = match ? match[1] : '';
		if (indent === undefined || lineIndent.length < indent.length) {
			indent = lineIndent;
		}
	}
	return indent ?? '';
}

function stripCommonIndent(text: string): { stripped: string; indent: string } {
	const lines = splitLines(text);
	const indent = getCommonIndent(lines);
	if (indent.length === 0) return { stripped: text, indent: '' };

	const stripped = lines
		.map((line) => (line.startsWith(indent) ? line.slice(indent.length) : line))
		.join('\n');
	return { stripped, indent };
}

function reIndent(text: string, indent: string): string {
	if (indent.length === 0) return text;
	return splitLines(text)
		.map((line) => (line.trim().length > 0 ? indent + line : line))
		.join('\n');
}

// ---------------------------------------------------------------------------
// Strategies
// ---------------------------------------------------------------------------

function exactMatch(
	content: string,
	oldStr: string,
	newStr: string,
): FuzzyMatchResult | null {
	const index = content.indexOf(oldStr);
	if (index === -1) return null;

	// Ensure unique match
	if (content.indexOf(oldStr, index + 1) !== -1) return null;

	return {
		replaced:
			content.slice(0, index) + newStr + content.slice(index + oldStr.length),
		strategy: 'exact',
	};
}

function lineTrimmedMatch(
	content: string,
	oldStr: string,
	newStr: string,
): FuzzyMatchResult | null {
	const contentLines = splitLines(content);
	const oldLines = splitLines(oldStr);
	const trimmedOld = oldLines.map((l) => l.trim());

	if (oldLines.length === 0) return null;

	// Slide window over content lines
	let matchStart = -1;
	let matchCount = 0;

	for (let i = 0; i <= contentLines.length - oldLines.length; i++) {
		let matches = true;
		for (let j = 0; j < oldLines.length; j++) {
			if (contentLines[i + j].trim() !== trimmedOld[j]) {
				matches = false;
				break;
			}
		}
		if (matches) {
			matchCount++;
			if (matchCount > 1) return null; // Not unique
			matchStart = i;
		}
	}

	if (matchStart === -1) return null;

	const before = contentLines.slice(0, matchStart);
	const after = contentLines.slice(matchStart + oldLines.length);
	const result = [...before, newStr, ...after].join('\n');

	return { replaced: result, strategy: 'line-trimmed' };
}

function whitespaceNormalizedMatch(
	content: string,
	oldStr: string,
	newStr: string,
): FuzzyMatchResult | null {
	const contentLines = splitLines(content);
	const oldLines = splitLines(oldStr);
	const normalizedOld = oldLines.map((l) => l.trim().replace(/\s+/g, ' '));

	if (oldLines.length === 0) return null;

	let matchStart = -1;
	let matchCount = 0;

	for (let i = 0; i <= contentLines.length - oldLines.length; i++) {
		let matches = true;
		for (let j = 0; j < oldLines.length; j++) {
			const normalizedContent = contentLines[i + j].trim().replace(/\s+/g, ' ');
			if (normalizedContent !== normalizedOld[j]) {
				matches = false;
				break;
			}
		}
		if (matches) {
			matchCount++;
			if (matchCount > 1) return null;
			matchStart = i;
		}
	}

	if (matchStart === -1) return null;

	const before = contentLines.slice(0, matchStart);
	const after = contentLines.slice(matchStart + oldLines.length);
	const result = [...before, newStr, ...after].join('\n');

	return { replaced: result, strategy: 'whitespace-normalized' };
}

function indentationFlexibleMatch(
	content: string,
	oldStr: string,
	newStr: string,
): FuzzyMatchResult | null {
	const contentLines = splitLines(content);
	const { stripped: strippedOld } = stripCommonIndent(oldStr);
	const oldLines = splitLines(strippedOld);

	if (oldLines.length === 0) return null;

	let matchStart = -1;
	let matchIndent = '';
	let matchCount = 0;

	for (let i = 0; i <= contentLines.length - oldLines.length; i++) {
		// Get the indent of the first non-empty matched line
		const blockLines = contentLines.slice(i, i + oldLines.length);
		const { stripped: strippedBlock, indent } = stripCommonIndent(
			blockLines.join('\n'),
		);

		if (splitLines(strippedBlock).join('\n') === oldLines.join('\n')) {
			matchCount++;
			if (matchCount > 1) return null;
			matchStart = i;
			matchIndent = indent;
		}
	}

	if (matchStart === -1) return null;

	const before = contentLines.slice(0, matchStart);
	const after = contentLines.slice(matchStart + oldLines.length);
	const reIndented = reIndent(newStr, matchIndent);
	const result = [...before, reIndented, ...after].join('\n');

	return { replaced: result, strategy: 'indentation-flexible' };
}

function blockAnchorLevenshteinMatch(
	content: string,
	oldStr: string,
	newStr: string,
): FuzzyMatchResult | null {
	const contentLines = splitLines(content);
	const oldLines = splitLines(oldStr);

	if (oldLines.length < 2) return null;

	const firstOldTrimmed = oldLines[0].trim();
	const lastOldTrimmed = oldLines[oldLines.length - 1].trim();

	if (firstOldTrimmed.length === 0 || lastOldTrimmed.length === 0) return null;

	const tolerance = 0.3;
	let matchStart = -1;
	let matchEnd = -1;
	let matchCount = 0;

	for (let i = 0; i < contentLines.length; i++) {
		if (contentLines[i].trim() !== firstOldTrimmed) continue;

		// Found first-line anchor; search for last-line anchor
		const maxEnd = Math.min(
			i + oldLines.length + Math.ceil(oldLines.length * 0.5),
			contentLines.length,
		);

		for (let j = i + oldLines.length - 1; j < maxEnd; j++) {
			if (contentLines[j].trim() !== lastOldTrimmed) continue;

			// Check interior via Levenshtein
			const candidateBlock = contentLines.slice(i, j + 1).join('\n');
			const dist = levenshtein(oldStr, candidateBlock);
			const maxLen = Math.max(oldStr.length, candidateBlock.length);

			if (maxLen > 0 && dist / maxLen <= tolerance) {
				matchCount++;
				if (matchCount > 1) return null;
				matchStart = i;
				matchEnd = j;
			}
		}
	}

	if (matchStart === -1 || matchEnd === -1) return null;

	const before = contentLines.slice(0, matchStart);
	const after = contentLines.slice(matchEnd + 1);
	const result = [...before, newStr, ...after].join('\n');

	return { replaced: result, strategy: 'block-anchor-levenshtein' };
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function fuzzyMatch(
	content: string,
	oldStr: string,
	newStr: string,
): FuzzyMatchResult | null {
	// Try strategies in order of strictness
	return (
		exactMatch(content, oldStr, newStr) ??
		lineTrimmedMatch(content, oldStr, newStr) ??
		whitespaceNormalizedMatch(content, oldStr, newStr) ??
		indentationFlexibleMatch(content, oldStr, newStr) ??
		blockAnchorLevenshteinMatch(content, oldStr, newStr)
	);
}

File: simse-core/src/tools/host/fuzzy_edit.rs
//! Fuzzy Edit — 5-strategy matching engine for file edits.
//!
//! Ports `src/ai/tools/host/fuzzy-edit.ts` to Rust.
//!
//! Strategies tried in order of strictness:
//! 1. Exact match
//! 2. Line-trimmed (trim each line, compare)
//! 3. Whitespace-normalized (collapse internal whitespace)
//! 4. Indentation-flexible (strip common indent, re-indent replacement)
//! 5. Block-anchor + Levenshtein (match first/last line, 30% tolerance)

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// Result of a successful fuzzy match.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FuzzyMatchResult {
	/// The content with the replacement applied.
	pub replaced: String,
	/// The strategy that succeeded.
	pub strategy: String,
}

// ---------------------------------------------------------------------------
// Levenshtein distance (two-row DP)
// ---------------------------------------------------------------------------

/// Compute the Levenshtein edit distance between two strings.
///
/// Uses a space-optimized two-row algorithm.
pub fn levenshtein(a: &str, b: &str) -> usize {
	let a_bytes = a.as_bytes();
	let b_bytes = b.as_bytes();
	let m = a_bytes.len();
	let n = b_bytes.len();

	if m == 0 {
		return n;
	}
	if n == 0 {
		return m;
	}

	let mut prev: Vec<usize> = (0..=n).collect();
	let mut curr: Vec<usize> = vec![0; n + 1];

	for i in 1..=m {
		curr[0] = i;
		for j in 1..=n {
			let cost = if a_bytes[i - 1] == b_bytes[j - 1] {
				0
			} else {
				1
			};
			curr[j] = (prev[j] + 1)
				.min(curr[j - 1] + 1)
				.min(prev[j - 1] + cost);
		}
		std::mem::swap(&mut prev, &mut curr);
	}

	prev[n]
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Split text into lines (by '\n').
fn split_lines(text: &str) -> Vec<&str> {
	text.split('\n').collect()
}

/// Get the common leading whitespace across non-empty lines.
fn get_common_indent(lines: &[&str]) -> String {
	let non_empty: Vec<&&str> = lines.iter().filter(|l| !l.trim().is_empty()).collect();
	if non_empty.is_empty() {
		return String::new();
	}

	// Compute the longest common prefix of leading whitespace across all non-empty lines
	let first = non_empty[0];
	let first_ws_len = first.len() - first.trim_start().len();
	let first_indent = &first[..first_ws_len];
	let mut prefix_len = first_indent.len();

	for line in &non_empty[1..] {
		let ws_len = line.len() - line.trim_start().len();
		let line_indent = &line[..ws_len];
		prefix_len = prefix_len.min(line_indent.len());
		// Compare byte-by-byte up to prefix_len
		for (i, (a, b)) in first_indent.bytes().zip(line_indent.bytes()).enumerate() {
			if i >= prefix_len {
				break;
			}
			if a != b {
				prefix_len = i;
				break;
			}
		}
	}

	first_indent[..prefix_len].to_string()
}

/// Strip the common indent from text, returning (stripped_text, indent_string).
fn strip_common_indent(text: &str) -> (String, String) {
	let lines = split_lines(text);
	let indent = get_common_indent(&lines);
	if indent.is_empty() {
		return (text.to_string(), String::new());
	}

	let stripped: Vec<&str> = lines
		.iter()
		.map(|line| {
			if line.starts_with(&indent) {
				&line[indent.len()..]
			} else {
				line
			}
		})
		.collect();

	(stripped.join("\n"), indent)
}

/// Re-indent text by prepending `indent` to every non-empty line.
fn re_indent(text: &str, indent: &str) -> String {
	if indent.is_empty() {
		return text.to_string();
	}
	split_lines(text)
		.iter()
		.map(|line| {
			if line.trim().is_empty() {
				(*line).to_string()
			} else {
				format!("{}{}", indent, line)
			}
		})
		.collect::<Vec<_>>()
		.join("\n")
}

// ---------------------------------------------------------------------------
// Strategy 1: Exact match
// ---------------------------------------------------------------------------

fn exact_match(content: &str, old_str: &str, new_str: &str) -> Option<FuzzyMatchResult> {
	let index = content.find(old_str)?;

	// Ensure unique match — check no second non-overlapping occurrence
	// Use index + old_str.len() to stay on UTF-8 boundary
	if content[index + old_str.len()..].contains(old_str) {
		return None;
	}

	let replaced =
		String::from(&content[..index]) + new_str + &content[index + old_str.len()..];

	Some(FuzzyMatchResult {
		replaced,
		strategy: "exact".to_string(),
	})
}

// ---------------------------------------------------------------------------
// Strategy 2: Line-trimmed match
// ---------------------------------------------------------------------------

fn line_trimmed_match(content: &str, old_str: &str, new_str: &str) -> Option<FuzzyMatchResult> {
	let content_lines = split_lines(content);
	let old_lines = split_lines(old_str);
	let trimmed_old: Vec<&str> = old_lines.iter().map(|l| l.trim()).collect();

	if old_lines.is_empty() {
		return None;
	}

	let mut match_start: Option<usize> = None;
	let mut match_count = 0;

	if content_lines.len() >= old_lines.len() {
		for i in 0..=(content_lines.len() - old_lines.len()) {
			let mut matches = true;
			for j in 0..old_lines.len() {
				if content_lines[i + j].trim() != trimmed_old[j] {
					matches = false;
					break;
				}
			}
			if matches {
				match_count += 1;
				if match_count > 1 {
					return None;
				}
				match_start = Some(i);
			}
		}
	}

	let start = match_start?;

	let mut result_parts: Vec<&str> = Vec::new();
	for line in &content_lines[..start] {
		result_parts.push(line);
	}
	result_parts.push(new_str);
	for line in &content_lines[start + old_lines.len()..] {
		result_parts.push(line);
	}

	Some(FuzzyMatchResult {
		replaced: result_parts.join("\n"),
		strategy: "line-trimmed".to_string(),
	})
}

// ---------------------------------------------------------------------------
// Strategy 3: Whitespace-normalized match
// ---------------------------------------------------------------------------

/// Collapse runs of whitespace into single spaces, then trim.
fn normalize_whitespace(s: &str) -> String {
	let trimmed = s.trim();
	let mut result = String::with_capacity(trimmed.len());
	let mut in_space = false;
	for ch in trimmed.chars() {
		if ch.is_whitespace() {
			if !in_space {
				result.push(' ');
				in_space = true;
			}
		} else {
			result.push(ch);
			in_space = false;
		}
	}
	result
}

fn whitespace_normalized_match(
	content: &str,
	old_str: &str,
	new_str: &str,
) -> Option<FuzzyMatchResult> {
	let content_lines = split_lines(content);
	let old_lines = split_lines(old_str);
	let normalized_old: Vec<String> = old_lines.iter().map(|l| normalize_whitespace(l)).collect();

	if old_lines.is_empty() {
		return None;
	}

	let mut match_start: Option<usize> = None;
	let mut match_count = 0;

	if content_lines.len() >= old_lines.len() {
		for i in 0..=(content_lines.len() - old_lines.len()) {
			let mut matches = true;
			for j in 0..old_lines.len() {
				if normalize_whitespace(content_lines[i + j]) != normalized_old[j] {
					matches = false;
					break;
				}
			}
			if matches {
				match_count += 1;
				if match_count > 1 {
					return None;
				}
				match_start = Some(i);
			}
		}
	}

	let start = match_start?;

	let mut result_parts: Vec<&str> = Vec::new();
	for line in &content_lines[..start] {
		result_parts.push(line);
	}
	result_parts.push(new_str);
	for line in &content_lines[start + old_lines.len()..] {
		result_parts.push(line);
	}

	Some(FuzzyMatchResult {
		replaced: result_parts.join("\n"),
		strategy: "whitespace-normalized".to_string(),
	})
}

// ---------------------------------------------------------------------------
// Strategy 4: Indentation-flexible match
// ---------------------------------------------------------------------------

fn indentation_flexible_match(
	content: &str,
	old_str: &str,
	new_str: &str,
) -> Option<FuzzyMatchResult> {
	let content_lines = split_lines(content);
	let (stripped_old, _) = strip_common_indent(old_str);
	let old_lines = split_lines(&stripped_old);

	if old_lines.is_empty() {
		return None;
	}

	let mut match_start: Option<usize> = None;
	let mut match_indent = String::new();
	let mut match_count = 0;

	if content_lines.len() >= old_lines.len() {
		for i in 0..=(content_lines.len() - old_lines.len()) {
			let block_lines: Vec<&str> = content_lines[i..i + old_lines.len()].to_vec();
			let block_text = block_lines.join("\n");
			let (stripped_block, indent) = strip_common_indent(&block_text);

			let stripped_block_lines = split_lines(&stripped_block);
			let old_lines_joined: Vec<&str> = old_lines.to_vec();

			if stripped_block_lines == old_lines_joined {
				match_count += 1;
				if match_count > 1 {
					return None;
				}
				match_start = Some(i);
				match_indent = indent;
			}
		}
	}

	let start = match_start?;

	let re_indented = re_indent(new_str, &match_indent);

	let before: Vec<&str> = content_lines[..start].to_vec();
	let after: Vec<&str> = content_lines[start + old_lines.len()..].to_vec();

	let mut result = before.join("\n");
	if !before.is_empty() {
		result.push('\n');
	}
	result.push_str(&re_indented);
	if !after.is_empty() {
		result.push('\n');
		result.push_str(&after.join("\n"));
	}

	Some(FuzzyMatchResult {
		replaced: result,
		strategy: "indentation-flexible".to_string(),
	})
}

// ---------------------------------------------------------------------------
// Strategy 5: Block-anchor + Levenshtein match
// ---------------------------------------------------------------------------

fn block_anchor_levenshtein_match(
	content: &str,
	old_str: &str,
	new_str: &str,
) -> Option<FuzzyMatchResult> {
	let content_lines = split_lines(content);
	let old_lines = split_lines(old_str);

	if old_lines.len() < 2 {
		return None;
	}

	let first_old_trimmed = old_lines[0].trim();
	let last_old_trimmed = old_lines[old_lines.len() - 1].trim();

	if first_old_trimmed.is_empty() || last_old_trimmed.is_empty() {
		return None;
	}

	let tolerance = 0.3_f64;
	let mut match_start: Option<usize> = None;
	let mut match_end: Option<usize> = None;
	let mut match_count = 0;

	for i in 0..content_lines.len() {
		if content_lines[i].trim() != first_old_trimmed {
			continue;
		}

		// Found first-line anchor; search for last-line anchor
		let max_end_exclusive = (i + old_lines.len()
			+ ((old_lines.len() as f64 * 0.5).ceil() as usize))
			.min(content_lines.len());

		let search_start = if !old_lines.is_empty() {
			i + old_lines.len() - 1
		} else {
			i
		};

		for j in search_start..max_end_exclusive {
			if content_lines[j].trim() != last_old_trimmed {
				continue;
			}

			// Check interior via Levenshtein
			let candidate_block = content_lines[i..=j].join("\n");
			let dist = levenshtein(old_str, &candidate_block);
			let max_len = old_str.len().max(candidate_block.len());

			if max_len > 0 && (dist as f64 / max_len as f64) <= tolerance {
				match_count += 1;
				if match_count > 1 {
					return None;
				}
				match_start = Some(i);
				match_end = Some(j);
			}
		}
	}

	let start = match_start?;
	let end = match_end?;

	let before: Vec<&str> = content_lines[..start].to_vec();
	let after: Vec<&str> = content_lines[end + 1..].to_vec();

	let mut result = before.join("\n");
	if !before.is_empty() {
		result.push('\n');
	}
	result.push_str(new_str);
	if !after.is_empty() {
		result.push('\n');
		result.push_str(&after.join("\n"));
	}

	Some(FuzzyMatchResult {
		replaced: result,
		strategy: "block-anchor-levenshtein".to_string(),
	})
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Try 5 matching strategies in order of strictness:
/// 1. Exact match — `content.contains(old_str)`, unique occurrence only
/// 2. Line-trimmed — trim each line, compare line-by-line, unique window only
/// 3. Whitespace-normalized — collapse internal whitespace, compare, unique only
/// 4. Indentation-flexible — strip common indent, match, re-indent replacement
/// 5. Block-anchor + Levenshtein — match first/last lines exactly (trimmed),
///    check interior via Levenshtein distance <= 30% tolerance
pub fn fuzzy_match(content: &str, old_str: &str, new_str: &str) -> Option<FuzzyMatchResult> {
	exact_match(content, old_str, new_str)
		.or_else(|| line_trimmed_match(content, old_str, new_str))
		.or_else(|| whitespace_normalized_match(content, old_str, new_str))
		.or_else(|| indentation_flexible_match(content, old_str, new_str))
		.or_else(|| block_anchor_levenshtein_match(content, old_str, new_str))
}
